src/garrysmod/gamemodes/zombiesurvival/gamemode/sv_database.lua

```diff
@@ -2,31 +2,130 @@ require "pg"
                 ErrorNoHalt("Failed to set connection encoding:\n", err)
             end
-
-            if isfunction(onConnected) then onConnected(self) end
+            
+            MsgN("Connected to the database!")
+            
+            self._Connected = true
+            hook.Run("DatabaseConnected")
         else
             ErrorNoHalt("Unable to connect to the database!\n", err)
+            hook.Run("DatabaseConnectionFailed", error_text)
         end
     else
         ErrorNoHalt("gmsv_pg has not been found.")
     end
 end
+
+function GM.Database:Connected()
+    if self._Connected then
+        local _,is_open = pcall(function() return self.Connection:is_open() end)
+
+        -- If a connection has already been established, we can check the status
+        -- via pqxx::connection::is_open. Otherwise, use pcall to see if the "pg
+        -- - no connection, connect to a database first." error has been thrown.
+        if isbool(is_open) then
+            return is_open
+        else
+            return false
+        end
+    else
+        return false
+    end
+end
+
+function GM.Database:Escape(str)
+    return self.Connection:escape(str)
+end
+
+function GM.Database:Quote(str)
+    return self.Connection:quote(str)
+end
+
+function GM.Database:QuoteName(str)
+    return self.Connection:quote_name(str)
+end
+
+function GM.Database:Queue(query, callback)
+    assert(type(query) == "string")
+    table.insert(self._Queue, { query, callback })
+end
+
+function GM.Database:RawQuery(query, callback)
+    if not self:Connected() then
+        return self:Queue(query, callback)
+    end
+    
+    local query_obj = self.Connection:query(query)
+    query_obj:set_sync(true)
+    
+    local success, res, size = query_obj:run()
+    if success then
+        if isfunction(callback) then return callback(res, size) end
+    else
+        ErrorNoHalt("PostgreSQL Query Error!\n")
+        ErrorNoHalt(string.format("Query: %s\n", query))
+        local err = tostring(res)
+        ErrorNoHalt(string.format("%s\n", err))
+
+        if err == "Connection to database failed" then
+            return self:Queue(query, callback)
+        end
+    end
+end
+
+function GM.Database:_Think()
+    if #self._Queue > 0 then
+        if istable(self._Queue[1]) then
+            local queue_obj = self._Queue[1]
+            local query_string = queue_obj[1]
+            
+            if isstring(query_string) then
+                self:RawQuery(query_string, queue_obj[2])
+            end
+            
+            table.remove(self._Queue, 1)
+        end
+    end
+end
+
+timer.Create("GAMEMODE::Database#_Think", 1, 0, function()
+    GAMEMODE.Database:_Think()
+end)
+
+function GM.Database:_CheckConnection()
+    local connected = self:Connected()
+    
+    if not connected then self:Connect() end
+    return connected
+end
+
+timer.Create("GAMEMODE::Database#_CheckConnection", 10, 0, function()
+    GAMEMODE.Database:_CheckConnection()
+end)
```

Comment:
This is really cool. Bring the database down in another tab.

```bash
docker-compose stop database
```

Then, queue a query up.

```lua
lua_run GAMEMODE.Database:Queue("SELECT version();", function(res,_) PrintTable(res) end)
```

You can inspect the contents of the queue with.

```bash
dump_postgresql_queue
1: SELECT version();
```

It won't run until you bring the database back up.

```bash
docker-compose up database
```

```bash
. . .
Connected to the database!
Query: SELECT version();
1:
                version =       PostgreSQL 12.2 (Debian 12.2-2.pgdg100+1) on x86_64-pc-linux-gnu, compiled by gcc (Debian 8.3.0-6) 8.3.0, 64-bit
```